rfcs/20180817-variables-20.md

```diff
@@ -0,0 +1,140 @@
+
+These are just two examples of how it's much harder to reason about TensorFlow programs when the value of a variable can change after it was read.
+
+Before resource handles TensorFlow variables were represented using a "ref" edge. A ref edge is a pair of pointers, one to a Tensor and one to a mutex, owned by something other than the tf runtime. When an op takes a ref tensor its input has to be a ref tensor, but when an op takes a non-ref tensor but its input is a ref tensor the pointer is silently dereferenced. This means that normal tensor objects in the graph can silently alias a mutable tensor, and hence two ops with the same input can see it having different values. Which value will be seen can depend on execution-specific details such as whether the variables are on a local or remote device, and in general it's not easy to ensure that a read happens before or after a specific write.
+
+
+### Internal resource variable ops
+
+We will expose the internal ops used to implement ResourceVariable as tf.experimental.variable_operations (name TBD). This way users and libraries can, if they need to, modify the behavior of variables at will.
+
+
+## Migration plan
+
+The migration plan is roughly as follows. TODO(apassos): flesh out this section with cost estimates.
+
+
+
+1.  Implement the abstract base class and factory function scope under the hood
+1.  Expose the factory function scope as tf.variable_creator_scope
+1.  LSC to change tf.variable_scope / tf.get_variable to tf.compat.v1.*
+1.  Removal of tf.variable_scope and tf.get_variable from the tf 2 namespace
+1.  Implement the subclass to be returned from tf.assign*
+1.  LSC to change tf.assign*(v, …) to v.assign*(...)
+1.  Change the implementation of tf.compat.v1.variable_scope to not rely on a per-graph variable scope stack
+1.  Remove the get_variable_scope and related public methods from tf.Graph (leaving them on tf.compat.v1.Graph)
+1.  Implement PartitionedVariable as a subclass of the tf.Variable interface
+1.  Add a partitioner scope to the tf 2.0 API
+1.  Add a deprecation warning to the tf.compat.v1 partitioned variable scope with a migration warning
+1.  [questionable] Implement a variable creator factory function which calls get_variable under the hood
+1.  Make this function active in all tf.compat.v1 endpoints which currently call get_variable (with a decorator, probably)
+1.  Change the behavior in tf2 to call tf.Variable (which will redirect to tf.get_variable in tf.compat.v1, keeping the existing behavior but cleaning the codebase)
+1.  [WARNING: checkpoint-breaking change] drop calls to variable_scope in parts of our API which use it. Right now they are: feature_column, rnn, canned estimators, optimizer slots, TPU estimator. Most can be replaced with judicious use of name= arguments
+1.  [optional] Implement tf v2 make_template which does not rely on variable_scope internally and uses a factory creator function to track and reuse variables
```

Comment:
Going to request that this is a requirement instead of an optional.

Reply:
Alternatively, expand layer-based APIs to make it easier to reuse existing variables imperatively.

Reply:
Also want to request `make_template` in v2.
One question: right now in `make_template`, one can use `get_variable` to create reused variable and `tf.Variable(trainable=False)` to create local (unshared) variables. After `get_variable` is deprecated I wonder what should be the alternative.

Reply:
@alextp could you comment on how `make_template` will be supported and how a user should create shared and unshared variables inside `make_template` ?